pkg/multicluster/v1/kubeconfigs_snapshot_emitter.sk.go

```diff
@@ -178,6 +197,165 @@ func (c *kubeconfigsEmitter) Snapshots(watchNamespaces []string, opts clients.Wa
+
+		namespacesResources, err := k.CoreV1().Namespaces().List(ctx, metav1.ListOptions{FieldSelector: excludeNamespacesFieldDesciptors})
+		if err != nil {
+			return nil, nil, err
+		}
+		allOtherNamespaces := make([]string, len(namespacesResources.Items))
+		for _, ns := range namespacesResources.Items {
+			allOtherNamespaces = append(allOtherNamespaces, ns.Namespace)
+		}
+
+		// nonWatchedNamespaces
+		// REFACTOR
+		for _, namespace := range allOtherNamespaces {
+			/* Setup namespaced watch for KubeConfig */
+			{
+				kubeconfigs, err := c.kubeConfig.List(namespace, clients.ListOpts{Ctx: opts.Ctx, ExpressionSelector: opts.ExpressionSelector})
+				if err != nil {
+					return nil, nil, errors.Wrapf(err, "initial KubeConfig list")
+				}
+				initialKubeConfigList = append(initialKubeConfigList, kubeconfigs...)
+				kubeconfigsByNamespace[namespace] = kubeconfigs
+			}
+			kubeConfigNamespacesChan, kubeConfigErrs, err := c.kubeConfig.Watch(namespace, opts)
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "starting KubeConfig watch")
+			}
+
+			done.Add(1)
+			go func(namespace string) {
+				defer done.Done()
+				errutils.AggregateErrs(ctx, errs, kubeConfigErrs, namespace+"-kubeconfigs")
+			}(namespace)
+			/* Watch for changes and update snapshot */
+			go func(namespace string) {
+				for {
+					select {
+					case <-ctx.Done():
+						return
+					case kubeConfigList, ok := <-kubeConfigNamespacesChan:
+						if !ok {
+							return
+						}
+						select {
+						case <-ctx.Done():
+							return
+						case kubeConfigChan <- kubeConfigListWithNamespace{list: kubeConfigList, namespace: namespace}:
+						}
+					}
+				}
+			}(namespace)
+		}
+		// create watch on all namespaces, so that we can add resources from new namespaces
+		namespaceWatch, err := k.CoreV1().Namespaces().Watch(opts.Ctx, metav1.ListOptions{FieldSelector: excludeNamespacesFieldDesciptors})
+		if err != nil {
+			return nil, nil, err
+		}
+
+		go func() {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case event, ok := <-namespaceWatch.ResultChan():
+					if !ok {
+						return
+					}
+					switch event.Type {
+					case kubewatch.Error:
+						errs <- errors.Errorf("receiving namespace event", event)
+					default:
+						namespacesResources, err := k.CoreV1().Namespaces().List(opts.Ctx, metav1.ListOptions{FieldSelector: excludeNamespacesFieldDesciptors})
+						if err != nil {
+							errs <- errors.Wrapf(err, "listing the namespace resources")
+						}
+
+						hit := false
+						newNamespaces := []string{}
+
+						for _, item := range namespacesResources.Items {
+							namespace := item.Namespace
+							_, hit = kubeconfigsByNamespace[namespace]
+							if !hit {
+								newNamespaces = append(newNamespaces, namespace)
+								continue
+							}
+						}
+						if hit {
+							// add a watch for all the new namespaces
+							// REFACTOR
+							for _, namespace := range newNamespaces {
```

Comment:
we don't need hit var.. if `newNamespaces` is empty list then we wont loop at all